### DISCUSSION TITLE
Программная реализация конкретных фукций в ZCAD

### DISCUSSION THREAD

**Discussion by gtfox** [zamtmn/zcad#128] [Q&A] [answered]
Для изучения внутреннего строения Zcad хотелось бы получить "пинок" в нужном направлении :) 
А именно: название файлов и процедур в них, которые отвечают за выполнение интересующих действий...

1) Поиск примитивов на чертеже и обработка данных их свойств (может в комплекте с передачей их в excel/csv)
2) Программное "рисование" на чертеже. Вставка примитивов строки спецификации/перечня элементов
3) Поиск/выбор примитивов в пределах области или другого примитива
4) Обработка события соединения проводов в стык с появлением точки в месте соединения

**Comment by zamtmn** [comment, marked as answer]
пока вкратце
движек https://github.com/zamtmn/zcad/tree/master/cad_source/zengine
зкад https://github.com/zamtmn/zcad/tree/master/cad_source/zcad
Зкад пишется давно и уже имел как успехи так и провалы. многие моменты нужно пересматривать-переделывать. код далек от идеала, в большинстве своем - говнокод

работа с чертежем организуется "командами" - это либо процедура, либо объект запускаемый из интерфейса для реализации некоторых действий, например ответом на пункт 2 будет команда LINE расположенная тут:
2 - https://github.com/zamtmn/zcad/blob/master/cad_source/zcad/commands/uzccommand_line.pas

поиск перебором примитивов чертежа тут
1 - https://github.com/zamtmn/zcad/blob/master/cad_source/zcad/commands/uzccommand_selectall.pas чертеж вперебирается 2 раза, первый раз подсчет примитивов, второй раз примитивы помечаются выбранными, причем если в чертеже примитивов немного - с отображением свойств в инспекторе

чертеж имеет древовидную структуру, есть функции для быстрого геометрического поиска. но пример мне нужно будет приготовить, с лету ничего путного не вспомнил
3 - ??

4 - провода... это пока в сыром варианте https://www.youtube.com/watch?v=y9K0Jgev9YI реализация
тут https://github.com/zamtmn/zcad/blob/master/cad_source/zcad/entities/uzcextdrschconnector.pas и https://github.com/zamtmn/zcad/blob/master/cad_source/zcad/entities/uzcextdrschconnection.pas соответственно расширение коннектора и соединительной линии этот вопрос лучше переговорить голосом

**Comment by gtfox** [comment]
Хотел бы немного уточнить свой запрос по первым 3-м пунктам:
1. Поиск примитива (по имени или параметру) и чтение/запись его полей данных
_selectall в этом плане не очень подходит, т.к. гребёт без разбора всё подряд :)
Тут подошёл бы экспорт перечня элементов. Там и выбор определённых примитивов и чтение их свойств
2. Программная вставка примитивов. Тут подошёл бы пример автосоздания однолинейных схем.
3. Поиск/выбор примитивов в пределах области или другого примитива. Здесь идеально подойдёт механизм "пользовательских областей" который показан на примере схемы внешних проводок, где автоматически проставляются номера клемм и проводов.
Я не смог найти видео где это показано...

**Comment by zamtmn** [comment]
1. то что ты имеешь ввиду жто команда "El_Material_Legend" но там смотреть нечего - написана кучу лет назад, много раз рефакторилась не глядя, чтоб сохранить работоспособность. Свойства примитивов есть как "запрограммированные" координаты, слой и т.п. так и "свободные" я из называю "переменные" доступ к переменным https://github.com/zamtmn/zcad/blob/master/cad_source/zcad/commands/uzccommand_examplevarmanipulation.pas
2. все команды рисующие стото сложное не годятся для знакомства - устарели смотри простые команды создания примитивов. ну и https://github.com/zamtmn/zcad/blob/master/cad_source/zcad/commands/uzccommand_exampleinsertdevice.pas
3. Этот механизм встроен в движек и работает сам по себе, на него никак не повлиять и не испльзовать, он просто делает доступными переменные обоасти для входящих в нее примитивов
вот поиск примитивов в точке https://github.com/zamtmn/zcad/blob/55c62e7614e2326a7d33d4fe9807b21605039897/cad_source/zcad/electrotech/uzcentelleader.pas#L303 выноска смотрит что есть там куда она указывает. также есть поиск по объему FindObjectsInVolume

**Comment by gtfox** [comment]
Имеется вопросы про контекстное меню чертежа/элемента чертежа

Из примера который мне попался я увидел, что меню формируется в редакторе
![Безымянный](https://github.com/user-attachments/assets/c8e9fd0a-416c-4c3c-af65-3c219f2bfdde)

1. Где найти контекстное меню чертежа/элемента чертежа ZCAD (к какой форме оно относится)?
2. Теоретический вопрос: добавление/удаление пунктов контекстного меню в зависимости от переменных устройства?

**Reply by Alligator-1** [reply]
Плотно взялся! 😁👍

**Comment by zamtmn** [comment]
Я не использую штатные визуально\формошлепательские возможности. всё это делается файлом описания экшенов, и файлом описания меню
https://github.com/zamtmn/zcad/blob/master/environment/runtimefiles/common/menu/actionscontent.xml
https://github.com/zamtmn/zcad/blob/master/environment/runtimefiles/zcadelectrotech/menu/menuscontent.xml
тут сложного ничего нет, но проще это обсудить голосом

**Comment by gtfox** [comment]
Оставлю тут для себя
Границы примитива .vp.BoundingBox
